File: tests/test_browser_webauthn_continuation.py
"""Offline adversarial tests for the private WebAuthn transition contract."""

import base64
import json
import os
import pathlib
import sys

import pytest

import file_utils

E2E_DIRECTORY = pathlib.Path(file_utils.get_repo_root()) / "tests" / "e2e"
sys.path.insert(0, str(E2E_DIRECTORY))

import e2e_browser_scenario_contract as browser_scenario_contract
import e2e_browser_webauthn_continuation as browser_webauthn


#============================================
def private_file(path: pathlib.Path, contents: str) -> None:
	"""Exclusively write one exact mode-0600 ASCII private test fixture."""
	file_descriptor = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
	with os.fdopen(file_descriptor, "wb") as output:
		output.write(contents.encode("ascii"))


#============================================
def continuation_contents(gateway_port: int = 55001) -> str:
	"""Return the closed, canonical resident-credential continuation fixture."""
	return json.dumps(
		{
			"version": 1,
			"origin": f"https://localhost:{gateway_port}",
			"rpId": "localhost",
			"credentials": [{
				"credentialId": "AA",
				"isResidentCredential": True,
				"rpId": "localhost",
				"privateKey": "AQI",
				"signCount": 0,
				"userHandle": "Aw",
				"backupEligibility": False,
				"backupState": False,
			}],
		},
		separators=(",", ":"),
	)


#============================================
def claimed_contract() -> browser_scenario_contract.ScenarioContract:
	"""Return the smallest claimed contract valid under the scenario boundary."""
	return browser_scenario_contract.ScenarioContract(
		scenario_id="claimed",
		spec_path="tests/playwright/e2e/claimed.spec.ts",
		personas=("morgan_sysadmin",),
		baseline_reads=("base_course",),
		ui_creates=("course",),
		sysadmin_requirement="claimed",
		visible_observation="visible_passkey_entry",
	)


#============================================
def acknowledgement_contents(
	scenario_id: str = "claimed",
	namespace: str = "bs1-0123456789ab-claimed",
	gateway_port: int = 55001,
) -> str:
	"""Return canonical evidence of a visible claimed Sysadmin passkey entry."""
	return json.dumps(
		{
			"event": "visible_sysadmin_passkey_sign_in",
			"namespace": namespace,
			"origin": f"https://localhost:{gateway_port}",
			"scenarioId": scenario_id,
			"schemaVersion": 1,
		},
		separators=(",", ":"),
	)


#============================================
def test_continuation_paths_are_private_and_claimed_acknowledgements_are_distinct(
	tmp_path: pathlib.Path,
) -> None:
	"""The owner selects fixed paths without letting a child choose either capability location."""
	private_directory = tmp_path.resolve()
	assert browser_webauthn.continuation_path(private_directory) == (
		private_directory / "webauthn-continuation.json"
	)
	assert browser_webauthn.acknowledgement_path(private_directory, "claimed") == (
		private_directory / "webauthn-continuation-ack-claimed.json"
	)
	assert browser_webauthn.acknowledgement_path(private_directory, "other") != (
		browser_webauthn.acknowledgement_path(private_directory, "claimed")
	)
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.acknowledgement_path(private_directory, "invalid-id!")


#============================================
def test_continuation_validator_rejects_adversarial_schema_metadata_and_user_handle(
	tmp_path: pathlib.Path,
) -> None:
	"""The owner accepts only canonical, bounded, current-user continuation data."""
	path = tmp_path / "webauthn-continuation.json"
	private_file(path, continuation_contents())
	browser_webauthn.validate_continuation(path, 55001)
	for change in (
		lambda value: value.update({"unexpected": True}),
		lambda value: value.update({"origin": "https://localhost:55002"}),
		lambda value: value.update({"rpId": "example.test"}),
		lambda value: value.update({"credentials": []}),
		lambda value: value.update({"credentials": value["credentials"] * 2}),
		lambda value: value["credentials"][0].update({"credentialId": "="}),
		lambda value: value["credentials"][0].update({"rpId": "example.test"}),
		lambda value: value["credentials"][0].update({"extra": True}),
		lambda value: value["credentials"][0].update({
			"userHandle": base64.urlsafe_b64encode(b"x" * 65).decode("ascii").rstrip("="),
		}),
	):
		value = json.loads(continuation_contents())
		change(value)
		path.unlink()
		private_file(path, json.dumps(value, separators=(",", ":")))
		with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
			browser_webauthn.validate_continuation(path, 55001)
	path.unlink()
	private_file(path, continuation_contents() + "\n")
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_continuation(path, 55001)
	path.chmod(0o644)
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_continuation(path, 55001)
	path.unlink()
	private_file(path, "x" * 16_385)
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_continuation(path, 55001)
	path.unlink()
	target = tmp_path / "continuation-target.json"
	private_file(target, continuation_contents())
	path.symlink_to(target)
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_continuation(path, 55001)


#============================================
def test_acknowledgement_validator_rejects_adversarial_metadata_and_bindings(
	tmp_path: pathlib.Path,
) -> None:
	"""Only a canonical post-passkey acknowledgement can consume a continuation."""
	contract = claimed_contract()
	namespace = "bs1-0123456789ab-claimed"
	path = tmp_path / "webauthn-continuation-ack-claimed.json"
	private_file(path, acknowledgement_contents())
	browser_webauthn.validate_acknowledgement(path, 55001, contract, namespace)
	for change in (
		lambda value: value.update({"unexpected": True}),
		lambda value: value.update({"event": "other"}),
		lambda value: value.update({"namespace": "bs1-0123456789ab-other"}),
		lambda value: value.update({"origin": "https://localhost:55002"}),
		lambda value: value.update({"scenarioId": "other"}),
		lambda value: value.update({"schemaVersion": 2}),
	):
		value = json.loads(acknowledgement_contents())
		change(value)
		path.unlink()
		private_file(path, json.dumps(value, separators=(",", ":")))
		with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
			browser_webauthn.validate_acknowledgement(path, 55001, contract, namespace)
	path.unlink()
	private_file(path, acknowledgement_contents() + "\n")
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_acknowledgement(path, 55001, contract, namespace)
	path.chmod(0o644)
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_acknowledgement(path, 55001, contract, namespace)
	path.unlink()
	private_file(path, "x" * 1_025)
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_acknowledgement(path, 55001, contract, namespace)
	path.unlink()
	target = tmp_path / "acknowledgement-target.json"
	private_file(target, acknowledgement_contents())
	path.symlink_to(target)
	with pytest.raises(browser_webauthn.BrowserWebAuthnContinuationError):
		browser_webauthn.validate_acknowledgement(path, 55001, contract, namespace)

File: tests/test_browser_webauthn_continuation.mjs
import assert from "node:assert/strict";
import { chmodSync, mkdtempSync, readFileSync, rmSync, symlinkSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import test from "node:test";

import {
  webAuthnContinuationAcknowledgementPathFromEnvironment,
  webAuthnContinuationFromEnvironment,
  webAuthnContinuationPathForProducerFromEnvironment,
} from "../tests/playwright/browser_suite_live_config.ts";
import {
  createVirtualAuthenticator,
  exportWebAuthnContinuation,
  importWebAuthnContinuationIntoSession,
  removeVirtualAuthenticator,
  writeWebAuthnContinuationAcknowledgement,
  writeWebAuthnContinuation,
} from "../tests/playwright/helper_live_demo.ts";

function continuation(overrides = {}) {
  return {
    version: 1,
    origin: "https://localhost:55123",
    rpId: "localhost",
    credentials: [
      {
        credentialId: "AA",
        isResidentCredential: true,
        rpId: "localhost",
        privateKey: "AQI",
        signCount: 0,
        userHandle: "Aw",
        backupEligibility: false,
        backupState: false,
      },
    ],
    ...overrides,
  };
}

function claimedScenarioInput(overrides = {}) {
  return {
    schemaVersion: 2,
    scenarioId: "auth_authorization",
    namespace: "bs1-0123456789ab-auth_authorization",
    baseUrl: "https://localhost:55123/",
    personas: ["morgan_sysadmin"],
    baselineReads: ["genetics_practice_course"],
    sysadminRequirement: "claimed",
    visibleObservation: "visible_passkey_entry",
    ...overrides,
  };
}

function parse(contents) {
  return webAuthnContinuationFromEnvironment(
    { PLE_BROWSER_SUITE_WEBAUTHN_CONTINUATION_FILE: "/private/continuation.json" },
    "https://localhost:55123",
    () => contents,
    () => {},
  );
}

function temporaryDirectory() {
  const directory = mkdtempSync(join(tmpdir(), "ple-webauthn-continuation-"));
  return {
    directory,
    cleanup() {
      rmSync(directory, { force: true, recursive: true });
    },
  };
}

test("WebAuthn continuation decoder accepts one canonical resident credential", () => {
  const parsed = parse(JSON.stringify(continuation()));
  assert.equal(parsed.origin, "https://localhost:55123");
  assert.equal(parsed.credentials.length, 1);
  assert.equal(parsed.credentials[0].credentialId, "AA");
});

test("WebAuthn continuation decoder rejects unsafe paths and adversarial private files", () => {
  assert.throws(() => webAuthnContinuationPathForProducerFromEnvironment({}));
  assert.throws(() =>
    webAuthnContinuationPathForProducerFromEnvironment({
      PLE_BROWSER_SUITE_WEBAUTHN_CONTINUATION_FILE: "relative.json",
    }),
  );
  assert.throws(() => webAuthnContinuationAcknowledgementPathFromEnvironment({}));
  assert.throws(() =>
    webAuthnContinuationAcknowledgementPathFromEnvironment({
      PLE_BROWSER_SUITE_WEBAUTHN_CONTINUATION_ACK_FILE: "relative.json",
    }),
  );
  assert.throws(() =>
    webAuthnContinuationPathForProducerFromEnvironment({
      PLE_BROWSER_SUITE_WEBAUTHN_CONTINUATION_FILE: " /private/continuation.json",
    }),
  );

  const temporary = temporaryDirectory();
  try {
    const input = join(temporary.directory, "continuation.json");
    const target = join(temporary.directory, "target.json");
    const environment = { PLE_BROWSER_SUITE_WEBAUTHN_CONTINUATION_FILE: input };
    assert.throws(() =>
      webAuthnContinuationFromEnvironment(environment, "https://localhost:55123"),
    );
    writeFileSync(target, JSON.stringify(continuation()), { encoding: "ascii", mode: 0o600 });
    symlinkSync(target, input);
    assert.throws(() =>
      webAuthnContinuationFromEnvironment(environment, "https://localhost:55123"),
    );
    rmSync(input);
    writeFileSync(input, JSON.stringify(continuation()), { encoding: "ascii", mode: 0o600 });
    chmodSync(input, 0o644);
    assert.throws(() =>
      webAuthnContinuationFromEnvironment(environment, "https://localhost:55123"),
    );
    chmodSync(input, 0o600);
    writeFileSync(input, "x".repeat(16_385), { encoding: "ascii", mode: 0o600 });
    assert.throws(() =>
      webAuthnContinuationFromEnvironment(environment, "https://localhost:55123"),
    );
  } finally {
    temporary.cleanup();
  }
});

test("WebAuthn continuation decoder rejects schema, canonical, origin, RP, and credential changes", () => {
  const base = continuation();
  const variants = [
    { ...base, unexpected: true },
    { ...base, origin: "https://localhost:55123/" },
    { ...base, origin: "https://localhost:55124" },
    { ...base, rpId: "example.test" },
    { ...base, credentials: [] },
    { ...base, credentials: [base.credentials[0], base.credentials[0]] },
    {
      ...base,
      credentials: [{ ...base.credentials[0], rpId: "example.test" }],
    },
    {
      ...base,
      credentials: [{ ...base.credentials[0], credentialId: "=" }],
    },
    {
      ...base,
      credentials: [
        { ...base.credentials[0], userHandle: Buffer.from("x".repeat(65)).toString("base64url") },
      ],
    },
    {
      ...base,
      credentials: [{ ...base.credentials[0], backupState: true }],
    },
    {
      ...base,
      credentials: [{ ...base.credentials[0], extra: true }],
    },
  ];
  for (const value of variants) assert.throws(() => parse(JSON.stringify(value)));
  assert.throws(() => parse(`${JSON.stringify(base)}\n`));
  assert.throws(() => parse(JSON.stringify({ origin: base.origin, ...base })));
});

test("WebAuthn continuation writer exclusively creates canonical mode-0600 evidence", () => {
  const temporary = temporaryDirectory();
  try {
    const path = join(temporary.directory, "continuation.json");
    writeWebAuthnContinuation(path, continuation());
    assert.equal(readFileSync(path, "ascii"), JSON.stringify(continuation()));
    assert.throws(() => writeWebAuthnContinuation(path, continuation()));
  } finally {
    temporary.cleanup();
  }
});

test("WebAuthn acknowledgement writer exclusively binds a claimed visible passkey entry", () => {
  const temporary = temporaryDirectory();
  try {
    const path = join(temporary.directory, "acknowledgement.json");
    const input = claimedScenarioInput();
    writeWebAuthnContinuationAcknowledgement(path, input);
    assert.equal(
      readFileSync(path, "ascii"),
      JSON.stringify({
        event: "visible_sysadmin_passkey_sign_in",
        namespace: input.namespace,
        origin: "https://localhost:55123",
        scenarioId: input.scenarioId,
        schemaVersion: 1,
      }),
    );
    assert.throws(() => writeWebAuthnContinuationAcknowledgement(path, input));
    assert.throws(() =>
      writeWebAuthnContinuationAcknowledgement(
        join(temporary.directory, "ordinary.json"),
        claimedScenarioInput({ sysadminRequirement: "not_required" }),
      ),
    );
  } finally {
    temporary.cleanup();
  }
});

class FakeCdpSession {
  constructor(authenticatorId, credentialResponse = undefined) {
    this.authenticatorId = authenticatorId;
    this.credentialResponse = credentialResponse;
    this.calls = [];
  }

  async send(method, params = undefined) {
    this.calls.push(params === undefined ? { method } : { method, params });
    if (method === "WebAuthn.addVirtualAuthenticator") {
      return { authenticatorId: this.authenticatorId };
    }
    if (method === "WebAuthn.getCredentials") return this.credentialResponse;
    if (method === "WebAuthn.addCredential") return {};
    return {};
  }
}

test("WebAuthn helper normalizes CDP binary fields and uses the exact export/import calls", async () => {
  const source = new FakeCdpSession("source-auth", {
    credentials: [
      {
        credentialId: "/w==",
        isResidentCredential: true,
        rpId: "localhost",
        privateKey: "AQI=",
        signCount: 9,
        userHandle: "Aw==",
        backupEligibility: false,
        backupState: false,
        userDisplayName: "Morgan Reyes",
        userName: "morgan.reyes@live-demo.ple.example",
      },
    ],
  });
  const sourceAuthenticator = await createVirtualAuthenticator(source);
  const exported = await exportWebAuthnContinuation(
    sourceAuthenticator,
    "https://localhost:55123",
    "localhost",
  );
  assert.deepEqual(exported, {
    version: 1,
    origin: "https://localhost:55123",
    rpId: "localhost",
    credentials: [
      {
        credentialId: "_w",
        isResidentCredential: true,
        rpId: "localhost",
        privateKey: "AQI",
        signCount: 9,
        userHandle: "Aw",
        backupEligibility: false,
        backupState: false,
      },
    ],
  });
  assert.deepEqual(source.calls, [
    { method: "WebAuthn.enable" },
    {
      method: "WebAuthn.addVirtualAuthenticator",
      params: {
        options: {
          protocol: "ctap2",
          transport: "internal",
          hasResidentKey: true,
          hasUserVerification: true,
          isUserVerified: true,
          automaticPresenceSimulation: true,
        },
      },
    },
    { method: "WebAuthn.getCredentials", params: { authenticatorId: "source-auth" } },
  ]);

  const target = new FakeCdpSession("target-auth");
  await importWebAuthnContinuationIntoSession(target, exported);
  assert.deepEqual(target.calls, [
    { method: "WebAuthn.enable" },
    {
      method: "WebAuthn.addVirtualAuthenticator",
      params: {
        options: {
          protocol: "ctap2",
          transport: "internal",
          hasResidentKey: true,
          hasUserVerification: true,
          isUserVerified: true,
          automaticPresenceSimulation: true,
        },
      },
    },
    {
      method: "WebAuthn.addCredential",
      params: {
        authenticatorId: "target-auth",
        credential: {
          credentialId: "/w==",
          isResidentCredential: true,
          rpId: "localhost",
          privateKey: "AQI=",
          signCount: 9,
          userHandle: "Aw==",
          backupEligibility: false,
          backupState: false,
        },
      },
    },
  ]);

  await removeVirtualAuthenticator(sourceAuthenticator);
  assert.deepEqual(source.calls.slice(-2), [
    {
      method: "WebAuthn.removeVirtualAuthenticator",
      params: { authenticatorId: "source-auth" },
    },
    { method: "WebAuthn.disable" },
  ]);
});

test("WebAuthn helper rejects a CDP record outside its closed resident-credential projection", async () => {
  const session = new FakeCdpSession("source-auth", {
    credentials: [
      {
        credentialId: "AA==",
        isResidentCredential: true,
        rpId: "localhost",
        privateKey: "AQI=",
        signCount: 0,
        userHandle: "Aw==",
        backupEligibility: false,
        backupState: false,
        userDisplayName: "Morgan Reyes",
        userName: "morgan.reyes@live-demo.ple.example",
        unexpected: true,
      },
    ],
  });
  const authenticator = await createVirtualAuthenticator(session);
  await assert.rejects(() =>
    exportWebAuthnContinuation(authenticator, "https://localhost:55123", "localhost"),
  );
});
